File: go/tests/hosting_multiplayer_test.go
// hosting_multiplayer_test.go
//
// Verifies CircleAI.Hosting.Multiplayer ports: GuestPeerIdentity defaults,
// MultiplayerHub join/leave/cursor/edit broadcasts through an injected
// broadcaster, LWW-by-rev edit acceptance/rejection, presence snapshots, and the
// ColourFor cursor-colour hash.

package circleai_test

import (
	"sync"
	"testing"

	circleai "github.com/bhengubv/CircleAI/go"
)

// captureBroadcaster records every SendToOthersInGroup call.
type captureBroadcaster struct {
	mu    sync.Mutex
	calls []broadcast
}

type broadcast struct {
	group  string
	except string
	method string
	args   []interface{}
}

func (b *captureBroadcaster) SendToOthersInGroup(group, except, method string, args ...interface{}) {
	b.mu.Lock()
	b.calls = append(b.calls, broadcast{group, except, method, args})
	b.mu.Unlock()
}

func (b *captureBroadcaster) methods() []string {
	b.mu.Lock()
	defer b.mu.Unlock()
	var m []string
	for _, c := range b.calls {
		m = append(m, c.method)
	}
	return m
}

func TestGuestPeerIdentity_Defaults(t *testing.T) {
	g := circleai.NewGuestPeerIdentity("", "")
	if g.DisplayName() != "Guest" {
		t.Errorf("display name = %q, want Guest", g.DisplayName())
	}
	if len(g.PeerID()) != 32 { // GUID hex without dashes
		t.Errorf("peerID length = %d, want 32", len(g.PeerID()))
	}
	g2 := circleai.NewGuestPeerIdentity("pid", "Alice")
	if g2.PeerID() != "pid" || g2.DisplayName() != "Alice" {
		t.Errorf("explicit values not honoured: %q / %q", g2.PeerID(), g2.DisplayName())
	}
}

func TestMultiplayerHub_JoinCursorEdit(t *testing.T) {
	circleai.MultiplayerResetStateForTesting()
	defer circleai.MultiplayerResetStateForTesting()

	bc := &captureBroadcaster{}
	hub, err := circleai.NewMultiplayerHub(circleai.NewGuestPeerIdentity("peer-1", "Alice"), bc)
	if err != nil {
		t.Fatalf("ctor: %v", err)
	}

	hub.OnConnected("conn-1")
	hub.JoinDocument("conn-1", "doc-A")

	peers := circleai.MultiplayerPeers("doc-A")
	if len(peers) != 1 || peers[0].DisplayName != "Alice" || peers[0].DocID != "doc-A" {
		t.Fatalf("presence wrong: %+v", peers)
	}

	hub.SendCursor("conn-1", "doc-A", 3, 7)

	// First edit at rev 5 is accepted.
	if got := hub.SendEdit("conn-1", "doc-A", "hello", 5); got != 5 {
		t.Errorf("rev5 accept returned %d, want 5", got)
	}
	if circleai.MultiplayerCurrentRev("doc-A") != 5 {
		t.Errorf("current rev = %d, want 5", circleai.MultiplayerCurrentRev("doc-A"))
	}

	// Strictly-stale edit at rev 2 (< current 5) is rejected → server returns the
	// current rev (5) and does NOT broadcast EditApplied (newRev.Rev != rev).
	if got := hub.SendEdit("conn-1", "doc-A", "stale", 2); got != 5 {
		t.Errorf("stale edit returned %d, want current rev 5", got)
	}

	// Newer edit at rev 9 accepted.
	if got := hub.SendEdit("conn-1", "doc-A", "newer", 9); got != 9 {
		t.Errorf("rev9 accept returned %d, want 9", got)
	}
	if circleai.MultiplayerCurrentRev("doc-A") != 9 {
		t.Errorf("current rev = %d, want 9", circleai.MultiplayerCurrentRev("doc-A"))
	}

	methods := bc.methods()
	// Expected broadcasts: PeerJoined, CursorChanged, EditApplied (rev5),
	// EditApplied (rev9). The strictly-stale rev2 does NOT broadcast.
	countEditApplied := 0
	for _, m := range methods {
		if m == "EditApplied" {
			countEditApplied++
		}
	}
	if countEditApplied != 2 {
		t.Errorf("EditApplied broadcasts = %d, want 2 (strictly-stale one suppressed): %v", countEditApplied, methods)
	}
}

func TestMultiplayerHub_Disconnect(t *testing.T) {
	circleai.MultiplayerResetStateForTesting()
	defer circleai.MultiplayerResetStateForTesting()

	bc := &captureBroadcaster{}
	hub, _ := circleai.NewMultiplayerHub(circleai.NewGuestPeerIdentity("p", "Bob"), bc)
	hub.OnConnected("c1")
	hub.JoinDocument("c1", "doc-X")
	hub.OnDisconnected("c1")

	if len(circleai.MultiplayerPeers("doc-X")) != 0 {
		t.Error("peer should be removed on disconnect")
	}
	found := false
	for _, m := range bc.methods() {
		if m == "PeerLeft" {
			found = true
		}
	}
	if !found {
		t.Error("PeerLeft not broadcast on disconnect")
	}
}

func TestMultiplayerHub_NilIdentityRejected(t *testing.T) {
	if _, err := circleai.NewMultiplayerHub(nil, nil); err == nil {
		t.Error("nil identity should error")
	}
}

func TestColourForPeer(t *testing.T) {
	// Empty id → fixed fallback.
	if circleai.ColourForPeer("") != "#5a4fcf" {
		t.Errorf("empty peer colour = %q", circleai.ColourForPeer(""))
	}
	// Deterministic + valid HSL for a non-empty id.
	c1 := circleai.ColourForPeer("alice")
	c2 := circleai.ColourForPeer("alice")
	if c1 != c2 {
		t.Error("colour must be deterministic")
	}
	if len(c1) < 4 || c1[:4] != "hsl(" {
		t.Errorf("colour = %q, want hsl(...)", c1)
	}
	// Different ids generally differ.
	if circleai.ColourForPeer("bob") == circleai.ColourForPeer("carol") {
		t.Log("hash collision (rare) — acceptable")
	}
}

File: rust/tests/hosting_multiplayer_test.rs
//! hosting_multiplayer_test.rs
//!
//! Verifies MultiplayerHub: connect/join/leave/disconnect presence, cursor +
//! edit broadcasts, LWW-by-rev edit resolution, per-doc rev tracking, and the
//! ColourFor hash→HSL. Mirrors the C# MultiplayerHub state machine.

use circle_ai::hosting_multiplayer::{
    colour_for, GuestPeerIdentity, HubBroadcast, IMultiplayerPeerIdentity, MultiplayerHub,
};

struct Fixed(&'static str, &'static str);
impl IMultiplayerPeerIdentity for Fixed {
    fn peer_id(&self) -> &str {
        self.0
    }
    fn display_name(&self) -> &str {
        self.1
    }
}

#[test]
fn guest_identity_defaults() {
    let g = GuestPeerIdentity::new(None, None);
    assert_eq!(g.display_name(), "Guest");
    assert_eq!(g.peer_id().len(), 32); // hex GUID, no dashes
}

#[test]
fn join_document_broadcasts_peer_joined() {
    let hub = MultiplayerHub::new();
    hub.on_connected("c1", &Fixed("peer-1", "Alice"));
    let bc = hub.join_document("c1", "doc-a").unwrap();
    match bc {
        HubBroadcast::PeerJoined {
            doc_id,
            connection_id,
            display_name,
            color,
        } => {
            assert_eq!(doc_id, "doc-a");
            assert_eq!(connection_id, "c1");
            assert_eq!(display_name, "Alice");
            assert_eq!(color, colour_for("peer-1"));
        }
        _ => panic!("expected PeerJoined"),
    }
    // Presence reflects the join.
    let peers = hub.peers("doc-a");
    assert_eq!(peers.len(), 1);
    assert_eq!(peers[0].display_name, "Alice");
}

#[test]
fn leave_document_broadcasts_peer_left_and_clears_presence() {
    let hub = MultiplayerHub::new();
    hub.on_connected("c1", &Fixed("p1", "Alice"));
    hub.join_document("c1", "doc-a");
    let bc = hub.leave_document("c1", "doc-a").unwrap();
    assert!(matches!(bc, HubBroadcast::PeerLeft { .. }));
    assert_eq!(hub.peers("doc-a").len(), 0);
}

#[test]
fn disconnect_while_in_doc_broadcasts_peer_left() {
    let hub = MultiplayerHub::new();
    hub.on_connected("c1", &Fixed("p1", "Alice"));
    hub.join_document("c1", "doc-a");
    let bc = hub.on_disconnected("c1").unwrap();
    assert!(matches!(bc, HubBroadcast::PeerLeft { .. }));
    assert_eq!(hub.peers("doc-a").len(), 0);
}

#[test]
fn disconnect_when_not_in_doc_returns_none() {
    let hub = MultiplayerHub::new();
    hub.on_connected("c1", &Fixed("p1", "Alice"));
    assert!(hub.on_disconnected("c1").is_none());
}

#[test]
fn send_cursor_broadcasts_position() {
    let hub = MultiplayerHub::new();
    hub.on_connected("c1", &Fixed("p1", "Alice"));
    let bc = hub.send_cursor("c1", "doc-a", 12, 5).unwrap();
    match bc {
        HubBroadcast::CursorChanged { line, ch, connection_id, .. } => {
            assert_eq!(line, 12);
            assert_eq!(ch, 5);
            assert_eq!(connection_id, "c1");
        }
        _ => panic!("expected CursorChanged"),
    }
}

#[test]
fn send_edit_accepts_higher_rev_and_broadcasts() {
    let hub = MultiplayerHub::new();
    hub.on_connected("c1", &Fixed("p1", "Alice"));
    let out = hub.send_edit("c1", "doc-a", "new content", 5);
    assert_eq!(out.accepted_rev, 5);
    match out.broadcast.unwrap() {
        HubBroadcast::EditApplied { content, rev, from_connection_id, doc_id } => {
            assert_eq!(content, "new content");
            assert_eq!(rev, 5);
            assert_eq!(from_connection_id, "c1");
            assert_eq!(doc_id, "doc-a");
        }
        _ => panic!("expected EditApplied"),
    }
    assert_eq!(hub.current_rev("doc-a"), 5);
}

#[test]
fn send_edit_rejects_stale_rev_no_broadcast() {
    let hub = MultiplayerHub::new();
    hub.on_connected("c1", &Fixed("p1", "Alice"));
    hub.send_edit("c1", "doc-a", "v10", 10);
    // A stale rev (7 <= 10) is rejected; caller gets the current rev back.
    let out = hub.send_edit("c1", "doc-a", "v7", 7);
    assert_eq!(out.accepted_rev, 10);
    assert!(out.broadcast.is_none());
    assert_eq!(hub.current_rev("doc-a"), 10);
}

#[test]
fn first_edit_clamps_rev_to_at_least_one() {
    let hub = MultiplayerHub::new();
    hub.on_connected("c1", &Fixed("p1", "Alice"));
    // rev 0 on a fresh doc clamps to max(0,1)=1 → the client's 0 is "stale".
    let out = hub.send_edit("c1", "doc-a", "content", 0);
    assert_eq!(out.accepted_rev, 1);
    assert!(out.broadcast.is_none());
    assert_eq!(hub.current_rev("doc-a"), 1);
}

#[test]
fn current_rev_zero_for_untouched_doc() {
    let hub = MultiplayerHub::new();
    assert_eq!(hub.current_rev("never"), 0);
}

#[test]
fn reset_state_clears_everything() {
    let hub = MultiplayerHub::new();
    hub.on_connected("c1", &Fixed("p1", "Alice"));
    hub.join_document("c1", "doc-a");
    hub.send_edit("c1", "doc-a", "x", 3);
    hub.reset_state_for_testing();
    assert_eq!(hub.peers("doc-a").len(), 0);
    assert_eq!(hub.current_rev("doc-a"), 0);
}

#[test]
fn colour_for_is_stable_and_themed() {
    // Empty id → the fixed fallback.
    assert_eq!(colour_for(""), "#5a4fcf");
    // Deterministic for the same id.
    assert_eq!(colour_for("peer-1"), colour_for("peer-1"));
    // HSL with fixed saturation + lightness.
    let c = colour_for("abc");
    assert!(c.starts_with("hsl("));
    assert!(c.ends_with(", 70%, 55%)"));
}

#[test]
fn colour_for_matches_csharp_hash_for_known_input() {
    // C#: h = 0; foreach c: h = h*31 + c;  hue = ((h % 360)+360)%360.
    // For "ab": h = (0*31 + 'a') = 97; then 97*31 + 'b'(98) = 3007 + 98 = 3105.
    // 3105 % 360 = 225. → "hsl(225, 70%, 55%)".
    assert_eq!(colour_for("ab"), "hsl(225, 70%, 55%)");
}
